components/scream/src/physics/p3/p3_functions.hpp

```diff
@@ -92,6 +92,115 @@ struct Functions
 
   using Workspace = typename WorkspaceManager<Spack, Device>::Workspace;
 
+  // This struct stores prognostic variables evolved by P3.
+  struct P3PrognosticState {
```

Comment:
How about putting these types in a separate header, say, `p3_data_structures.hpp`? It might help compartmentalize the code a bit more. There's a clear dependency: interfaces require data structures, but not viceversa.

Reply:
Yeah, I guess I stuck things in here to get the right typing information for the views in the struct from the template parameters of the `Functions` "namespace" struct. Further, these structs are only ever used in conjunction with `p3_main`--I don't think we'd ever use them without calling that specific function. So it seemed expedient to place them alongside it.

That said, I don't have a strong opinion about where these structs live. How do others feel about this? Should we punt on this for now and discuss this in the context of Luca's proposed restructuring (#478)?

Reply:
Yes, they will have a limited sets of customers: some high level p3 function, and atmosphere_microphysics.cpp. I just think it makes it easy to track down if they are isolated to a single file. But again, not a big deal (by any means), given the limited scope of the structs. So unless someone else agrees that putting them in a separate headers improves readability/maintainability, then they can stay where they are.

Reply:
And yes, maybe a discussion about this in the issue you linked might be more appropriate.